csharp/src/Ice/Endpoint.cs

```diff
@@ -196,11 +200,10 @@ public override string ToString()
         public abstract Connection CreateDatagramServerConnection(ObjectAdapter adapter);
 
         /// <summary>Expands endpoint into separate endpoints for each IP address returned by the DNS resolver.
-        /// </summary>
-        /// <returns>A value task holding the expanded endpoints if Host is a DNS name or IP address; otherwise, a value
-        /// task holding an empty collection.</returns>
+        /// Precondition: <see cref="HasDnsHost"/> is true.</summary>
+        /// <returns>A value task holding the expanded endpoints.</returns>
         protected internal virtual ValueTask<IEnumerable<Endpoint>> ExpandHostAsync(CancellationToken cancel) =>
-            new(ImmutableArray<Endpoint>.Empty);
+            throw new NotImplementedException();
```

Comment:
Would `InvalidOperationException` be better here?

Reply:
I think `NotImplementedException` is slightly better. You would get this exception if you override HasDnsHost to return true (sometimes) and forget to override ExpandHostAsync.